### PR TITLE
Refactor sampling

### PR DESCRIPTION
**Pull request by jgersak**
refactor expand_to_grids, sample_by_values
add test for get_mask_of_samples
refactor the rest of the testsy

### CODE REVIEW

File: ml_tools/eolearn/ml_tools/sampling.py [hunk 1]
```diff
@@ -60,33 +60,19 @@ def sample_by_values(
...
-    all_rows, all_columns = [], []
     for value, n_samples in n_samples_per_value.items():
-        n_points = value_counts.get(value, 0)
-
-        if n_samples > n_points and not replace:
-            raise ValueError(
-                f"For label {value} the requested amount of samples is {n_samples} but only {n_points} points are "
-                "available."
-            )
+        sample_indices = rng.choice(np.nonzero(image == value), size=n_samples, replace=replace, axis=1)
```
Comment (zigaLuksic) [inline ml_tools/eolearn/ml_tools/sampling.py:71]:
```
        sample_rows, sample_cols = rng.choice(np.nonzero(image == value), size=n_samples, replace=replace, axis=1)
```
consider this

File: ml_tools/eolearn/ml_tools/sampling.py [hunk 2]
```diff
@@ -60,33 +60,19 @@ def sample_by_values(
...
+        rows = np.concatenate((rows, sample_indices[0]))
+        columns = np.concatenate((columns, sample_indices[1]))
```
Comment (zigaLuksic) [inline ml_tools/eolearn/ml_tools/sampling.py:73]: now we're doing N (which is number of values) concatenations (which are `O(n)`), before we did N appends (which are O(1)) and one concatenation. It does simplify the code a bit i guess. Think about it a bit more

File: ml_tools/eolearn/tests/test_sampling.py [hunk 3]
```diff
@@ -92,82 +93,107 @@ def test_sample_by_values_errors(image: np.ndarray, n_samples: Dict[int, int]) -
...
-    sampled_uniques, sampled_counts = np.unique(eop.data["SAMPLED_DATA"], return_counts=True)
-    masked = eop.mask_timeless["sampling_mask"].squeeze(axis=2) == 1
-    masked_uniques, masked_counts = np.unique(eop.data["bands"][:, masked, :], return_counts=True)
-    assert_array_equal(sampled_uniques, masked_uniques, err_msg="Sampling mask not correctly describing sampled points")
-    assert_array_equal(sampled_counts, masked_counts, err_msg="Sampling mask not correctly describing sampled points")
 
+@pytest.mark.parametrize("seed", range(5))
+@pytest.mark.parametrize("block_task", [100, 5231, 0.4, 0], indirect=True)
+def test_object_sampling_task(eopatch: EOPatch, seed: int, block_task: BlockSamplingTask) -> None:
+    block_task.execute(eopatch, seed=seed)
```
Comment (zigaLuksic) [inline ml_tools/eolearn/tests/test_sampling.py:177]: i would include that in the previous test

File: ml_tools/eolearn/tests/test_sampling.py [hunk 4]
```diff
@@ -184,28 +210,31 @@ def test_fraction_sampling_errors(fraction: Union[float, Dict[int, float]], repl
...
+        fraction=request.param[0],
+        exclude_values=request.param[1],
+    )
+
+
 @pytest.mark.parametrize("seed", range(3))
 @pytest.mark.parametrize(
-    "fraction, exclude",
+    "fraction_task",
     [
```
Comment (zigaLuksic) [inline ml_tools/eolearn/tests/test_sampling.py:227]:
```suggestion
    [  # fraction, exclude_values
```

just to know what the two parameters are

File: ml_tools/eolearn/tests/test_sampling.py [hunk 5]
```diff
@@ -215,77 +244,107 @@ def test_fraction_sampling(
...
     masked_uniques, masked_counts = np.unique(eopatch.data["NDVI"][:, masked], return_counts=True)
-    assert_array_equal(sampled_uniques, masked_uniques, err_msg="Sampling mask not correctly describing sampled points")
-    assert_array_equal(sampled_counts, masked_counts, err_msg="Sampling mask not correctly describing sampled points")
+    assert_array_equal(sampled_uniques, masked_uniques)
+    assert_array_equal(sampled_counts, masked_counts)
+
+
+@pytest.mark.parametrize("seed", range(3))
+@pytest.mark.parametrize("fraction_task", [[0.2, None], [0.4, [0, 1]]], indirect=True)
+def test_fraction_sampling(fraction_task: FractionSamplingTask, seed: int, test_eopatch: EOPatch) -> None:
```
Comment (zigaLuksic) [inline ml_tools/eolearn/tests/test_sampling.py:253]: change the name of this and `test_fraction_sampling_fine_tuned` to something like `test_fraction_sampling_input_fraction` and `test_fraction_sampling_input_dict`

File: ml_tools/eolearn/ml_tools/sampling.py [hunk 6]
```diff
@@ -60,33 +60,19 @@ def sample_by_values(
...
         points.
     """
-    rng = rng or np.random.default_rng()
     if image.ndim != 2:
         raise ValueError(f"Given image has shape {image.shape} but sampling operates only on 2D images")
 
-    uniques, counts = np.unique(image, return_counts=True)
-    value_counts = dict(zip(uniques, counts))
+    rng = rng or np.random.default_rng()
+    rows = np.empty([0], dtype=int)
```
[the remark this answers is not in the document]
Comment (jgersak) [reply ml_tools/eolearn/ml_tools/sampling.py]: done

File: ml_tools/eolearn/ml_tools/sampling.py (hunk 1 above)
Comment (jgersak) [reply ml_tools/eolearn/ml_tools/sampling.py]: considered, accepted, done

File: ml_tools/eolearn/tests/test_sampling.py (hunk 3 above)
Comment (jgersak) [reply ml_tools/eolearn/tests/test_sampling.py]: done

File: ml_tools/eolearn/tests/test_sampling.py (hunk 5 above)
Comment (jgersak) [reply ml_tools/eolearn/tests/test_sampling.py]: done

**Review by zigaLuksic — APPROVED** [review]

**Pull request merged**